keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java

```diff
@@ -71,20 +72,28 @@ public static void checkTensorsMatchNonScalarShapeOrAreScalar(int[] proposalShap
     }
 
     public static int[] checkAllShapesMatch(int[]... shapes) {
-        return checkAllShapesMatch(Arrays.stream(shapes));
+        return checkAllShapesMatch(Arrays.stream(shapes), Optional.empty());
+    }
+
+    public static int[] checkAllShapesMatch(String errorMessage, int[]... shapes) {
+        return checkAllShapesMatch(Arrays.stream(shapes), Optional.of(errorMessage));
+    }
+
+    public static int[] checkAllShapesMatch(String errorMessage, Collection<int[]> shapes) {
+        return checkAllShapesMatch(shapes.stream(), Optional.of(errorMessage));
     }
 
     public static int[] checkAllShapesMatch(Collection<int[]> shapes) {
-        return checkAllShapesMatch(shapes.stream());
+        return checkAllShapesMatch(shapes.stream(), Optional.empty());
     }
 
-    private static int[] checkAllShapesMatch(Stream<int[]> shapesStream) {
+    private static int[] checkAllShapesMatch(Stream<int[]> shapesStream, Optional<String> errorMessage) {
```

Comment:
Do we need this Optional method arguments? Can we remove the logic branch upstream? (e.g. two methods where one takes a String and the other doesnt and the one that doesn't calls the ones that does with a default message)